src/System.Management.Automation/help/HelpCommands.cs

```diff
@@ -733,37 +733,46 @@ internal static void VerifyParameterForbiddenInRemoteRunspace(Cmdlet cmdlet, str
     /// </summary>
     public static class GetHelpCodeMethods
     {
-        /// <summary>
-        /// Verifies if the InitialSessionState of the current process.
-        /// </summary>
-        /// <returns></returns>
-        private static bool DoesCurrentRunspaceIncludeCoreHelpCmdlet()
+        private static bool TrySingle<TSource>(IEnumerable<TSource> source, out TSource element)
         {
-            InitialSessionState iss =
-                System.Management.Automation.Runspaces.Runspace.DefaultRunspace.InitialSessionState;
-            if (iss != null)
-            {
-                IEnumerable<SessionStateCommandEntry> publicGetHelpEntries = iss
-                    .Commands["Get-Help"]
-                    .Where(entry => entry.Visibility == SessionStateEntryVisibility.Public);
-                if (publicGetHelpEntries.Count() != 1)
-                {
-                    return false;
-                }
-
-                foreach (SessionStateCommandEntry getHelpEntry in publicGetHelpEntries)
+            using (IEnumerator<TSource> e = source.GetEnumerator())
+            {
+                if (e.MoveNext())
                 {
-                    SessionStateCmdletEntry getHelpCmdlet = getHelpEntry as SessionStateCmdletEntry;
-                    if ((getHelpCmdlet != null) && (getHelpCmdlet.ImplementingType.Equals(typeof(GetHelpCommand))))
+                    TSource result = e.Current;
+                    if (!e.MoveNext())
                     {
+                        element = result;
                         return true;
                     }
                 }
             }
 
+            element = default;
             return false;
         }
 
+        /// <summary>
+        /// Verifies if the InitialSessionState of the current process.
+        /// </summary>
+        /// <returns></returns>
+        private static bool DoesCurrentRunspaceIncludeCoreHelpCmdlet()
+        {
+            InitialSessionState iss = Runspace.DefaultRunspace.InitialSessionState;
+            if (iss is null)
+            {
+                return false;
+            }
+
+            IEnumerable<SessionStateCommandEntry> publicGetHelpEntries = iss
+                .Commands["Get-Help"]
+                .Where(entry => entry.Visibility == SessionStateEntryVisibility.Public);
+
+            return TrySingle(publicGetHelpEntries, out SessionStateCommandEntry entry)
+                && entry is SessionStateCmdletEntry getHelpCmdlet
+                && getHelpCmdlet.ImplementingType.Equals(typeof(GetHelpCommand));
```

Comment:
> The new code takes the first entry in `publicGetHelpEntries` and emits it if the conditions are satisfied.

@rjmholt I'm sorry, I don't understand you here. The new code returns `true` if there is one entry in `publicGetHelpEntries`, and that entry satisfies the conditions.

Reply:
Looking at the code again, I see the `Count() == 1` check that guarantees the equivalence of the two implementations. So I'm wrong about that.

But in both implementations, the code doesn't offer much in the way of readability. After the refactor, we've spread the logic across more methods and we still have the `.Where()` closure to deal with.

Something like this would be much more readable and without us making any real sacrifice in the way of performance:

```csharp
/// <summary>
/// Checks whether the default runspace associated with the current thread has the standard Get-Help cmdlet.
/// <summary>
/// <return>True if Get-Help is found, false otherwise.</return>
private static bool CurrentRunspaceIncludesCoreHelpCmdlet()
{
    InitialSessionState iss = Runspace.DefaultRunspace.InitialSessionState;
    if (iss is null)
    {
        return false;
    }

    SessionStateCommandEntry getHelpEntry = null;
    foreach (SessionStateCommandEntry commandEntry in iss.Command["Get-Help"])
    {
        if (commandEntry.Visibility != SessionStateEntryVisibility.Public)
        {
            continue;
        }

        // If we have multiple entries for Get-Help,
        // our assumption is that the standard Get-Help is not available.
        if (getHelpEntry != null)
        {
            return false;
        }

        getHelpEntry = commandEntry;
    }

    return getHelpEntry is SessionStateCmdletEntry getHelpCmdlet
        && getHelpCmdlet.ImplementingType == typeof(GetHelpCommand);
}
```

Reply:
@rjmholt As an aside, if `TrySingle` made it into LINQ (https://github.com/dotnet/runtime/issues/16362), we could write the following: 

```csharp
private static bool DoesCurrentRunspaceIncludeCoreHelpCmdlet()
    => Runspace.DefaultRunspace.InitialSessionState is not null
        && Runspace.DefaultRunspace.InitialSessionState.Commands["Get-Help"]
            .Where(entry => entry.Visibility == SessionStateEntryVisibility.Public)
            .TrySingle(out SessionStateCommandEntry entry)
        && entry is SessionStateCmdletEntry getHelpCmdlet
        && getHelpCmdlet.ImplementingType.Equals(typeof(GetHelpCommand));
```

...but wasn't it LINQ style we were trying to avoid? 😜

On a serious note, your suggested implementation avoids the LINQ Count method and is more readable than the original code and all of my proposed implementations, so let's go with that.

Lastly, thanks for making detailed suggestions in your code reviews, I am paying attention, and going forward, I will try to refocus my contributions.

Reply:
I went for a `for` loop to avoid allocating an iterator.